### PR TITLE
fixed #1261, overflow when calculating work

### LINKED ISSUE #1261: Overflow when mining
steps to reproduce

1. change `difficulty` in `null_morden.json` to `0x1`
2. change `networkID` in `null_morde.json` (to mine on private chain)
3. start `ethminer`
4. start `parity`

  ```
cargo run --release -- --chain ./ethcore/res/null_morden.json
  ```

5. parity crashes

```
     Running `target/release/parity --chain ./ethcore/res/null_morden.json`
2016-06-13 13:52:55 Using a conversion rate of Ξ1 = US$16.53 (14403825000 wei/gas)
2016-06-13 13:52:55 Starting Parity/v1.2.0-unstable-b33482c-20160613/x86_64-macos/rustc1.9.0
2016-06-13 13:52:55 Configured for Morden using "NullEngine" engine
2016-06-13 13:52:55 Public node URL: enode://ea136c184e8b3789511bc64d13697d5f618c09f04045cc4db660ef05c077270c3180b369361dad528bb13503bf981d6522591d999f1fa2610bd6899ab4479ec3@0.0.0.0:30303
thread 'hyper-server' panicked at 'Overflow', util/bigint/src/uint.rs:1328
stack backtrace:
   1:        0x1035de578 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:        0x1035e1c55 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:        0x1035e188e - std::panicking::default_hook::haac48fa641db8fa2
   4:        0x1035cb3f6 - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:        0x10334e684 - std::sys_common::unwind::begin_unwind::hd78480d5ec5c51b2
   6:        0x103352e27 - _<uint..U256 as std..convert..From<uint..U512>>::from::h4668fd66c63735f4
   7:        0x1030f26d7 - ethcore::ethereum::ethash::Ethash::difficulty_to_boundary::h45c0f2fc53ae3ae9
   8:        0x102dbc903 - _<jsonrpc_core..io..DelegateMethod<T, F> as jsonrpc_core..commander..MethodCommand>::execute::hdaf8d72b11646dc7
   9:        0x1030772e6 - jsonrpc_core::commander::Commander::execute_method::h4fb0b35382cd4696
  10:        0x103079932 - jsonrpc_core::request_handler::RequestHandler::handle_call::h37f2a25b8542193d
  11:        0x103078a36 - jsonrpc_core::request_handler::RequestHandler::handle_request::h7814b19ff4301e05
  12:        0x10309fdea - jsonrpc_core::io::IoHandler::handle_request::h684b5a26bb611cd2
  13:        0x1030181d2 - _<ServerHandler as hyper..server..Handler<hyper..net..HttpStream>>::on_request_readable::h87cd010e3ee2fc84
  14:        0x103037178 - _<hyper..http..conn..Conn<K, T, H>>::read::hc5ee8846330209d8
  15:        0x1030377f0 - _<hyper..http..conn..Conn<K, T, H>>::read::hc5ee8846330209d8
  16:        0x103037dc4 - _<hyper..http..conn..Conn<K, T, H>>::read::hc5ee8846330209d8
  17:        0x103036f23 - _<hyper..http..conn..Conn<K, T, H>>::on_readable::hc9c48e14eb7d079f
  18:        0x10303440a - _<hyper..http..conn..Conn<K, T, H>>::ready::h9fad96c326b0d252
  19:        0x10302c4df - _<mio..event_loop..EventLoop<H>>::run_once::hb98c76d8402133cf
  20:        0x1030262c5 - _<hyper..Server<A>>::handle::_$u7b$$u7b$closure$u7d$$u7d$::hd38644bb0dba5d6d
  21:        0x103021c5f - std::sys_common::unwind::try::try_fn::h6f378f1c44ca94d5
  22:        0x1035dd7db - __rust_try
  23:        0x1035dd763 - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  24:        0x1030279e8 - _<F as std..boxed..FnBox<A>>::call_box::hcbf8e7155ff7fd14
  25:        0x1035e0de8 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  26:     0x7fff96f8299c - _pthread_body
  27:     0x7fff96f82919 - _pthread_start
2016-06-13 13:52:56 Finishing work, please wait...
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Closed', ../src/libcore/result.rs:746
stack backtrace:
   1:        0x1035de578 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:        0x1035e1c55 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:        0x1035e188e - std::panicking::default_hook::haac48fa641db8fa2
   4:        0x1035cb3f6 - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:        0x1035cc27e - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:        0x1035dd847 - rust_begin_unwind
   7:        0x103608c00 - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:        0x103054f99 - core::result::unwrap_failed::h4ffbb4ae06989851
   9:        0x103054edf - hyper::server::Listening::close::h08ed0aabe12182f0
  10:        0x1030455fe - _<Server as std..ops..Drop>::drop::h8522a191fc2475d9
  11:        0x102d1547b - std..option..Option<ethcore_rpc..Server>::drop.23489::h73386ff0c942fa6c
  12:        0x102d0824a - parity::execute_client::h4ea25e4f35a5a6b8
  13:        0x102cf59eb - parity::execute::h4228f2b3aef67c46
  14:        0x102cf2b6f - parity::main::hc0b8b9a4c77213f8
  15:        0x1035e1482 - std::sys_common::unwind::try::try_fn::h09ba69fd13531e58
  16:        0x1035dd7db - __rust_try
  17:        0x1035e12c4 - std::rt::lang_start::h5b0863080165c75e
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Any', ../src/libcore/result.rs:746
stack backtrace:
   1:        0x1035de578 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:        0x1035e1c55 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:        0x1035e188e - std::panicking::default_hook::haac48fa641db8fa2
   4:        0x1035cb3f6 - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:        0x1035cc27e - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:        0x1035dd847 - rust_begin_unwind
   7:        0x103608c00 - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:        0x103054ca0 - core::result::unwrap_failed::h0bdd56634916bdf6
   9:        0x103054b02 - _<server..Listening as std..ops..Drop>::drop::hf7eefae0ab712359
  10:        0x103054ef8 - hyper::server::Listening::close::h08ed0aabe12182f0
  11:        0x1030455fe - _<Server as std..ops..Drop>::drop::h8522a191fc2475d9
  12:        0x102d1547b - std..option..Option<ethcore_rpc..Server>::drop.23489::h73386ff0c942fa6c
  13:        0x102d0824a - parity::execute_client::h4ea25e4f35a5a6b8
  14:        0x102cf59eb - parity::execute::h4228f2b3aef67c46
  15:        0x102cf2b6f - parity::main::hc0b8b9a4c77213f8
  16:        0x1035e1482 - std::sys_common::unwind::try::try_fn::h09ba69fd13531e58
  17:        0x1035dd7db - __rust_try
  18:        0x1035e12c4 - std::rt::lang_start::h5b0863080165c75e
thread panicked while panicking. aborting.
error: An unknown error occurred

To learn more, run the command again with --verbose.
```

### PR DESCRIPTION
**Pull request by debris**
geth does it similarly https://github.com/ethereum/go-ethereum/blob/506c9277911746dfbab0a585aee736bd3095f206/miner/remote_agent.go#L112-L127

### CODE REVIEW

File: ethcore/src/ethereum/ethash.rs [hunk 1]
```diff
@@ -509,5 +510,11 @@ mod tests {
 		}
 	}
 
+	#[test]
+	fn test_difficulty_to_boundary() {
+		let _ = Ethash::difficulty_to_boundary(&U256::from(0));
```
Comment (tomusdrw) [inline ethcore/src/ethereum/ethash.rs]: Why not asserting the actuall result?

**Comment by gavofyork** [comment]
waiting for the answer...

File: ethcore/src/ethereum/ethash.rs [hunk 2]
```diff
@@ -255,12 +255,14 @@ impl Ethash {
 
 	/// Convert an Ethash boundary to its original difficulty. Basically just `f(x) = 2^256 / x`.
 	pub fn boundary_to_difficulty(boundary: &H256) -> U256 {
-		U256::from((U512::one() << 256) / U256::from(boundary.as_slice()).into())
+		let d =	cmp::max(U256::from(boundary), U256::one());
+		((U256::one() << 255) / d) << 1
```
Comment (gavofyork) [inline ethcore/src/ethereum/ethash.rs]: will still overflow if `d == U256::one()`

File: ethcore/src/ethereum/ethash.rs [hunk 3]
```diff
@@ -509,5 +510,14 @@ mod tests {
 		}
 	}
 
+	#[test]
+	fn test_difficulty_to_boundary() {
+		assert_eq!(Ethash::difficulty_to_boundary(&U256::from(0)), H256::from(0));
+		assert_eq!(Ethash::difficulty_to_boundary(&U256::from(1)), H256::from(0));
```
Comment (gavofyork) [inline ethcore/src/ethereum/ethash.rs]: condition is wrong - right side should really be `2^256`, though we might better approximate that in 256 bits with `2^256 - 1`.

File: ethcore/src/ethereum/ethash.rs [hunk 4]
```diff
@@ -509,5 +510,14 @@ mod tests {
 		}
 	}
 
+	#[test]
+	fn test_difficulty_to_boundary() {
+		assert_eq!(Ethash::difficulty_to_boundary(&U256::from(0)), H256::from(0));
```
Comment (gavofyork) [inline ethcore/src/ethereum/ethash.rs]: this function is undefined for 0 difficulty. see YP(160).

**Pull request merged**